SourceCode/raytracer/CRTTypes.h

```diff
@@ -245,5 +245,75 @@ class Triangle {
         return "Triangle: {" + v0.toString() + ", " + v1.toString() + ", " + v2.toString() + "}";
     }
 
+    /*
+    * Input: ray, rayProj
+    * Input: rayProj: projection of ray.direction onto camera direction
+    * Output: t: distance from ray.origin to the intersection point
+    * Output: p: intersection point
+    * return: true if the ray intersects the triangle
+    */
+    bool intersect(const Ray& ray, float& t, Vec3& p) const {
+        // Here assuming counter-clockwise order
+        Vec3 e0 = v1 - v0;
+        Vec3 e1 = v2 - v0;
+        Vec3 tri_normal = cross(e0, e1).normalize();
+
+        float rayProj = dot(ray.direction, tri_normal);
+
+        // if triangle is facing the ray, compute ray-plane intersection
+        if (rayProj < -1e-6) {
+            float rpDist = dot(tri_normal, ray.origin - v0);
+            t = rpDist / rayProj;
```

Comment:
The main problem is here - you are calculating the _inverse_ distance, because you are missing a `-`.
`t = -rpDist / rayProj` is the correct formula.